### DISCUSSION TITLE
paopao网关配置不成功。。。

### DISCUSSION THREAD

**Discussion by delete7456** [kkkgo/PaoPaoGateWay#13] [问和答]
我的配置
   主网关：192.168.2.1
   paopaodns：192.168.2.20
   paopaogateway：192.168.2.5

问题：在使用时翻墙不成功
使用nslookup查看
```shell
C:\Users\imtxc>nslookup youtube.com
服务器:  UnKnown
Address:  192.168.2.20

名称:    youtube.com
Address:  7.0.0.15
``` 
我的理解是dns已成功解析到了网关发出的fakeip地址，我也已经做好了静态路由表，但依旧无法打开网页。
但是如果直接将paopao网关作为默认网关，则正常全局翻墙。
ps：请原谅我简陋的网络知识。

下面是部分配置
paopaodns docker-compose.yaml
```shell
version: "3"

services:
  paopaodns:
    image: sliamb/paopaodns:latest
    container_name: PaoPaoDNS
    restart: always
    volumes:
      - ./data:/data
    environment:
      - TZ=Asia/Shanghai
      - UPDATE=weekly
      - DNS_SERVERNAME=PaoPaoDNS,blog.03k.org
      - DNSPORT=53
      - CNAUTO=yes
      - CNFALL=yes
      - CN_TRACKER=yes
      - USE_HOSTS=no
      - IPV6=no
      - SOCKS5=192.168.2.5:1080
      - SERVER_IP=192.168.2.5
      - CUSTOM_FORWARD=192.168.2.5:53
      - AUTO_FORWARD=yes
      - AUTO_FORWARD_CHECK=yes
      - HTTP_FILE=yes
    ports:
      - "53:53/udp"
      - "53:53/tcp"
      - "5304:5304/udp"
      - "5304:5304/tcp"
      - "7889:7889/tcp"
``` 

paopaoGateWay
ppgw.ini
```shell
#paopao-gateway

# mode=socks5|yaml|suburl|free
# default: free
mode=yaml

# Set fakeip's CIDR here
# default: fake_cidr=7.0.0.0/8
fake_cidr=7.0.0.0/8

# Set your trusted DNS here
# default: dns_ip=1.0.0.1
dns_ip=192.168.2.20
# default: dns_port=53
# If used with PaoPaoDNS, you can set the 5304 port
dns_port=5304

# Clash's web dashboard
clash_web_port="80"
clash_web_password="clashpass"

# default：openport=no
# socks+http mixed 1080
openport=yes

# default: udp_enable=no
udp_enable=no

# default:30
sleeptime=30

# socks5 mode settting
# default: socks5_ip=gatewayIP
socks5_ip="192.168.2.200"
# default: socks5_port="7890"
socks5_port="7890"

# yaml mode settting
# The yaml file in the same directory as the ppgw.ini.
# default: yamlfile=custom.yaml
yamlfile="custom.yaml"


# yaml and subrul mode setting
# fast_node=check/yes/no
fast_node=yes
test_node_url="https://www.youtube.com/generate_204"
ext_node="Traffic|Expire| GB|Days|Date"
```

**Comment by kkkgo** [comment]
你可以跟踪一下路由，看看你的请求是否能到达网关吗？

**Comment by kkkgo** [comment]
```
1. ping代理的域名看看是否ping通。
ping youtube.com
2. 跟踪被代理域名的路由（如果是linux，使用traceroute）。
tracert youtube.com
```

**Comment by kkkgo** [comment]
另外，如果有主路由的型号以及实际的接线拓扑更好。

**Reply by delete7456** [reply]
```shell
C:\Users\imtxc>tracert www.youtube.com

通过最多 30 个跃点跟踪
到 www.youtube.com [7.0.0.32] 的路由:

  1     1 ms     1 ms     1 ms  192.168.2.1
  2     2 ms     2 ms     2 ms  192.168.2.1
  3     6 ms     5 ms     5 ms  192.168.2.1
  4     7 ms     7 ms     6 ms  192.168.2.1
  5     6 ms     6 ms     6 ms  192.168.2.1
  6     9 ms     9 ms     7 ms  192.168.2.1
  7     8 ms     8 ms     9 ms  192.168.2.1
  8    26 ms    14 ms    12 ms  192.168.2.1
  9    11 ms    10 ms     9 ms  192.168.2.1
 10    17 ms    44 ms    12 ms  192.168.2.1
``` 
看来是静态路由没有配置好？
我的主路由是梅林380固件的6300v2，下面是静态路由的配置图
![image](https://github.com/kkkgo/PaoPaoGateWay/assets/107162126/c85048a0-270d-4e51-99e4-6e1b530107f2)

接线拓扑图
![网络拓扑图](https://github.com/kkkgo/PaoPaoGateWay/assets/107162126/8ec912c9-3648-4ccf-9348-9a4e5c26b005)
因为被错误整崩溃了（，所以先用了hyper-v做实验

**Comment by kkkgo** [comment]
看起来就是你的梅林主路由没有正确静态路由到ppgw。如果你正确路由，你应该能看到类似的结果：
![image](https://github.com/kkkgo/PaoPaoGateWay/assets/4530897/0e720eb9-429d-48e0-85e9-2e2770565f99)

另外，你的“虚拟交换机”是真实存在的交换机吗？还是其实都接在梅林路由器上。有部分路由器系统的静态路由存在bug，如果是接在主路由下的同一个局域网设备没法正常静态路由，单独接一个交换机出来又没问题。不过我看你的路由跟踪应该不属于这种情况。你可以ssh到主路由，使用`ip r`命令查看路由表。

**Reply by delete7456** [reply]
```shell
admin@R6300V2-F2A7:/tmp/home/root# ip r
10.99.128.1 dev ppp0  proto kernel  scope link 
192.168.2.0/24 dev br0  proto kernel  scope link  src 192.168.2.1 
169.254.0.0/16 dev eth0  proto kernel  scope link  src 169.254.219.255 
7.0.0.0/8 via 192.168.2.5 dev br0  metric 2 
127.0.0.0/8 dev lo  scope link 
default via 10.99.128.1 dev ppp0 
``` 
看起来静态路由是正确的
虚拟交换机是hyper-v的虚拟交换机，并不是接在梅林上面（其实是我拓扑图画错了，交换机也是hyper-v的一部分

**Reply by kkkgo** [reply]
你在梅林ssh上跟踪路由看看？（linux，使用traceroute）

**Reply by delete7456** [reply]
```shell
admin@R6300V2-F2A7:/tmp/home/root# traceroute www.youtube.com
traceroute to www.youtube.com (172.217.160.110), 30 hops max, 38 byte packets
 1  10.99.128.1 (10.99.128.1)  2.882 ms  3.126 ms  2.367 ms
 2  111.23.186.125 (111.23.186.125)  6.223 ms  111.23.186.121 (111.23.186.121)  2.649 ms  111.23.186.125 (111.23.186.125)  1.962 ms
 3  111.23.49.197 (111.23.49.197)  14.121 ms  9.807 ms  183.214.233.229 (183.214.233.229)  7.976 ms
 4  111.24.9.185 (111.24.9.185)  7.537 ms  111.24.9.169 (111.24.9.169)  9.872 ms  13.761 ms
 5  111.24.4.53 (111.24.4.53)  27.365 ms  111.24.3.233 (111.24.3.233)  28.387 ms  221.183.109.149 (221.183.109.149)  29.136 ms
 6  221.183.89.37 (221.183.89.37)  27.072 ms  221.183.89.5 (221.183.89.5)  31.666 ms  *
 7  *  *  *
 8  *  *^C

这个是没有跳至paopao网关？
```

**Reply by kkkgo** [reply]
你直接跟踪7.0.0.15就可以了。梅林上的DNS不一样不能用域名来跟踪

**Reply by Leno709394** [reply]
你这静态明显没跳到paopao网关。 给你个建议。你虚拟机都开起来了。再搞多个爱快或者ros添加静态。 两个paopao不动，试一遍，如果正常，就是你梅林问题。。r6300v2 当ap用就好。以爱快或ros为主

**Reply by Leno709394** [reply]
C:\Users\Administrator>tracert www.youtube.com

通过最多 30 个跃点跟踪
到 www.youtube.com [7.0.0.115] 的路由:

  1    <1 毫秒   <1 毫秒   <1 毫秒 192.168.168.253
  2    <1 毫秒   <1 毫秒   <1 毫秒 7.0.0.115

跟踪完成。

我主路由爱快。得到结果这样的

**Reply by delete7456** [reply]
```shell
admin@R6300V2-F2A7:/tmp/home/root# traceroute 7.0.0.15
traceroute to 7.0.0.15 (7.0.0.15), 30 hops max, 38 byte packets
 1  *  *  *
 2  *  *  *
 3  *  *  *
 4  *  *  *
 5  *  *^C

``` 
确实没有跳成功

**Reply by delete7456** [reply]
还想请教一下另一个问题，之前发现主路由会默认下发ipv6dns，导致电脑没有使用paopaodns，openwrt和梅林都有这种情况，请问爱快或者ros能解决这个问题吗？

**Reply by Leno709394** [reply]
你早说这个就解决问题了。不能用ipv6 dns。 用了就优先ipv6 dns解析了。 爱快 ros我试了都不行

**Reply by Leno709394** [reply]
ipv6 也没啥用。目前主要跑pcdn对v6要求多，pcdn用途不需要v6 dns也能用。我就开了ipv6给网心云跑。  不自动分配ipv6 dns。这样可以正常用泡泡。网心云又能连上ipv6

**Reply by kkkgo** [reply]
爱快应该是有ipv6 dns下发的开关的。openwrt当然也能设置不下发ipv6 dns。梅林就不清楚了，身边没有设备。另外PaoPaoDNS可以单独设置IPV6开关和模式：
>IPV6： 仅在CNAUTO=yes时生效，是否返回IPv6的解析结果，默认为no，如果没有IPv6环境，选择no可以节省内存。设置为yes返回IPv6的查询（为分流优化，非大陆双栈域名仅返回A记录）。如果设置为only6，则只对IPv6 only的域名返回IPv6结果（该项设置不影响force_cn_list.txt）。如果设置为raw，则不对IPv6结果做任何处理，直接返回原始记录。

不过宽带v6地址会有过期的问题，不可用的时候优先IPV6地址可能会导致上网体验卡顿。所以用不上的一般不建议开启或者使用`only6`。

**Reply by kkkgo** [reply]
ROS的关闭IPV6 DNS下发相关选项也许可以参考：https://www.cnblogs.com/itfat/p/10992289.html   
其他的应该也大同小异。

**Comment by Leno709394** [comment]
> 爱快应该是有ipv6 dns下发的开关的。openwrt当然也能设置不下发ipv6 dns。梅林就不清楚了，身边没有设备。另外PaoPaoDNS可以单独设置IPV6开关和模式：
> 
> > IPV6： 仅在CNAUTO=yes时生效，是否返回IPv6的解析结果，默认为no，如果没有IPv6环境，选择no可以节省内存。设置为yes返回IPv6的查询（为分流优化，非大陆双栈域名仅返回A记录）。如果设置为only6，则只对IPv6 only的域名返回IPv6结果（该项设置不影响force_cn_list.txt）。如果设置为raw，则不对IPv6结果做任何处理，直接返回原始记录。
> 
> 不过宽带v6地址会有过期的问题，不可用的时候优先IPV6地址可能会导致上网体验卡顿。所以用不上的一般不建议开启或者使用`only6`。

我开过yes了。不生效。客户端那里只要有v6 dns。不出5分钟。又全用v6 解析。不走v4 dns返回fakeip了

**Reply by kkkgo** [reply]
这个IPV6的选项的意思是paopaodns是否提供域名的ipv6解析结果（AAAA记录）。客户端选择用什么DNS，paopaodns肯定无法控制的。但返回双栈结果的话，客户端会优先使用v6结果，主要是避免这个情况。

**Reply by Leno709394** [reply]
那具体要怎么设置才对。泡泡dns设置v6我开了yes。  我客户端dns v4 分配的 paopaodns ip。    客户端dns v6 我试过手动设paopaodns的v6 地址进去，或者留空。  ping v6网站域名是不通的。   设公共v6 dns。就会导致paopaodns 失效。  因为会变成v6直接返回google的v6地址过来了

**Reply by kkkgo** [reply]
v6 dns应该为空，客户端才会优先选择使用paopadns提供的结果。ping v6网站域名不通，那你可以尝试检查下是否能正常返回v6的解析结果。比如(192.168.1.2为paopaodns)：
```
nslookup 6.ipw.cn 192.168.1.2
```

**Reply by Leno709394** [reply]
C:\Users\Administrator>nslookup 6.ipw.cn 192.168.168.252
服务器:  UnKnown
Address:  192.168.168.252

*** 没有 6.ipw.cn 可以使用的 internal type for both IPv4 and IPv6 Addresses (A+AAAA)记录

C:\Users\Administrator>nslookup 6.ipw.cn 192.168.168.253
服务器:  iKuaiOS
Address:  192.168.168.253

名称:    6.ipw.cn
Address:  2402:4e00:1013:e500:0:9671:f018:4947


我爱快是253   paopao是252  ，   爱快可以返回。  泡泡不行。 paopaodns自己的dns我也设成paopao的。跟这个有关么

**Reply by kkkgo** [reply]
没有关系。paopaodns只会用自己的解析方案，不会使用外部的dns设置。这个问题你可以在paopaodns的docker里面跑`debug.sh`，到PaoPaoDNS的issue或者discussions贴出调试日志来分析。

**Comment by TwoBall** [comment]
我居然解析到了youtube的ipv6地址。
```
C:\Users\Administrator>tracert youtube.com

通过最多 30 个跃点跟踪
到 youtube.com [2a03:2880:f127:283:face:b00c:0:25de] 的路由:

  1    <1 毫秒   <1 毫秒   <1 毫秒 240e:369:c932:8301::1001
  2     2 ms     1 ms     2 ms  240e:369:c900::1
  3     1 ms     1 ms     1 ms  240e:d:8201:825::2
  4     2 ms     2 ms     2 ms  240e:d:8201:f001::2
  5     7 ms     *        7 ms  240e:d:8001:f012::2
  6    21 ms    21 ms    21 ms  240e::1:21:91:7202
  7     *        *        *     请求超时。
  8     *        *        *     请求超时。
  9     *        *        *     请求超时。
```

**Reply by kkkgo** [reply]
你`nslookup youtube.com`一下看看，确认是不是在用PaoPaoDNS的解析。

**Comment by lesnolie** [comment]
你好，看你也用hyper-v，请问HyperV运行paopaogw，连接以后是这样的吗？
![image](https://github.com/kkkgo/PaoPaoGateWay/assets/81410185/1deb4f1f-1b58-4b61-88cc-17fb6236ddea)
